smt/pkg/smt/smt_batch.go

```diff
@@ -143,27 +137,52 @@ func (s *SMT) InsertBatch(ctx context.Context, logPrefix string, nodeKeys []*uti
 			maxInsertingNodePathLevel = insertingNodePathLevel
 		}
 	}
+	progressChan <- uint64(1)
+	stopProgressPrinter()
 
 	s.updateDepth(maxInsertingNodePathLevel)
 
+	totalDeleteOps := 0
+	for _, mapLevel0 := range nodeHashesForDelete {
+		for _, mapLevel1 := range mapLevel0 {
+			for _, mapLevel2 := range mapLevel1 {
+				totalDeleteOps += len(mapLevel2)
+			}
+		}
+	}
+
+	progressChanDel, stopProgressPrinterDel := zk.ProgressPrinter(fmt.Sprintf("[%s] SMT incremental progress (deletes)", logPrefix), uint64(totalDeleteOps), false)
 	for _, mapLevel0 := range nodeHashesForDelete {
 		for _, mapLevel1 := range mapLevel0 {
 			for _, mapLevel2 := range mapLevel1 {
 				for _, nodeHash := range mapLevel2 {
 					s.Db.DeleteByNodeKey(*nodeHash)
 					s.Db.DeleteHashKey(*nodeHash)
+					progressChanDel <- uint64(1)
 				}
 			}
 		}
 	}
+	stopProgressPrinterDel()
+
+	totalFinalizeOps := 0
```

Comment:
Let's use len(nodeValues) as a totalFInalizeOps. Then we can send 1 to the channel on each iteration rather than on each non-zero iteration. This way we will remove an entire loop.